weaviate/embedded.py

```diff
@@ -235,21 +265,84 @@ def start(self) -> None:
         print(f"Started {self.options.binary_path}: process ID {self.process.pid}")
         self.wait_till_listening()
 
-    def stop(self) -> None:
-        if self.process is not None:
+
+class Embedded(_EmbeddedBase):
+    def is_listening(self) -> bool:
+        up = self.__is_listening()
+        return up[0] and up[1]
+
+    def __is_listening(self) -> Tuple[bool, bool]:
+        http_listening, grpc_listening = False, False
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
             try:
-                self.process.terminate()
-                self.process.wait()
-            except ProcessLookupError:
-                print(
-                    f"""Tried to stop embedded weaviate process {self.process.pid}. Process was not found. So not doing
-                    anything"""
-                )
-            self.process = None
+                s.connect((self.options.hostname, self.options.port))
+                http_listening = True
+            except (socket.error, ConnectionRefusedError):
+                pass
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            try:
+                s.connect((self.options.hostname, self.grpc_port))
+                grpc_listening = True
+            except (socket.error, ConnectionRefusedError):
+                pass
+        return (http_listening, grpc_listening)
 
-    def ensure_running(self) -> None:
-        if not self.is_listening():
-            print(
-                f"Embedded weaviate wasn't listening on port {self.options.port}, so starting embedded weaviate again"
+    def start(self) -> None:
```

Comment:
most of this also seems identical to the other version - can we move the identical parts to the base class?

Reply:
Good point, will do!